File: June 2025/03.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution {
  public:
    int atMostK(string& s, int k) {
        unordered_map<char, int> freq;
        int left = 0, total = 0;

        for (int right = 0; right < s.size(); ++right) {
            freq[s[right]]++;

            while (freq.size() > k) {
                freq[s[left]]--;
                if (freq[s[left]] == 0) {
                    freq.erase(s[left]);
                }
                left++;
            }

            total += right - left + 1;
        }

        return total;
    }

    int countSubstr(string& s, int k) {
        return atMostK(s, k) - atMostK(s, k - 1);
    }
};

File: June 2025/03.py
from collections import defaultdict

class Solution:
    def countSubstr(self, s, k):
        def atMostK(s, k):
            count = defaultdict(int)
            left = 0
            total = 0
            
            for right in range(len(s)):
                count[s[right]] += 1
                
                while len(count) > k:
                    count[s[left]] -= 1
                    if count[s[left]] == 0:
                        del count[s[left]]
                    left += 1
                
                total += right - left + 1
            
            return total
        
        return atMostK(s, k) - atMostK(s, k - 1)
